[Networking/Windows] Added IPv6 support

gpotter2 opened the pull request:
> Well this is https://github.com/secdev/scapy/pull/380 and https://github.com/secdev/scapy/pull/397 but a way more cleaner...

**IT IS FINALLY WORKING**

So as today, IPv6 routes are not supported by scapy on Windows. Here is the IPv6 routes support on Windows 😃 

This patch:

- Fixes some route6.py functions to work on Windows
- Removes double warning logs when loading scapy (when winpcap is not installed)
- Add regression.uts tests

What it returns:
http://pastebin.com/hUU6ykcc

I've tested it: it's possible to send IPv6 packets

gpotter2 commented:
@guedou @p-l- here's my updated version of it...

gpotter2 commented on scapy/route6.py:
```diff
@@ -265,7 +266,7 @@ def route(self, dst, dev=None):
         k = dst
         if dev is not None:
-            k = dst + "%%" + dev
+            k = dst + "%%" + (dev.pcap_name if not isinstance(dev, basestring) else dev)
```
It's dev.pcap_name right ? Cause i wasn't sure if it shouldn't have been dev.name...

p-l- replied:
Cannot remember... @guedou?
BTW, you should reverse the test: `dev if isinstance(dev, basestring) else dev.pcap_name`

gpotter2 replied:
Sure, I'll wait for his answer before pushing it...

gpotter2 commented on scapy/route6.py:
```diff
@@ -200,7 +201,7 @@ def route(self, dst, dev=None):
         k = dst
         if dev is not None:
-            k = dst + "%%" + dev
+            k = dst + "%%" + (dev if isinstance(dev, basestring) else dev.pcap_name)
```
It's dev.pcap_name right ? Cause i wasn't sure if it shouldn't have been dev.name...

> p-l-: Cannot remember... @guedou?

guedou replied:
I don't know either.

gpotter2 replied:
Oh we don't care: we just need that both read cache and set cache methods use the same thing... I will leave pcap_name

guedou commented on scapy/arch/pcapdnet.py:
```diff
@@ -37,10 +37,11 @@ def winpcapy_get_if_list():
   except OSError as e:
     if conf.interactive:
-      log_loading.error("Unable to import libpcap library: %s" % e)
+      log_loading.warning("wpcap.dll is not installed. You won't be able to send/recieve packets. Visit https://www.winpcap.org/ to install it")
```
I think that it is better to refer to Scapy documentation, especially because wpcap.dll could be installed differently.

guedou commented on scapy/arch/winpcapy.py:
```diff
@@ -28,7 +28,7 @@
     _lib_name = find_library('pcap')
     if not _lib_name:
-      raise OSError("Cannot fine libpcap.so library")
+        raise OSError("Cannot fine libpcap.so library")
```
I would rather keep winpcapy.py as it is. It would be easier to compare this included file to other ones.

guedou commented:
Thanks for this work. It looks amazing.

Would you mind splitting this PR into smaller ones ? That will make the fix easier to review, and later one easier to refer to. I prefer many commits with small changes. On the long term, they are easier to maintain. During development, it is also nicer as one fix does not break the checks.

For example, you could do several PRs, such as: no wpcap.dll, read_routes6 & mocked tests, double warnings, ...

gpotter2 commented:
I understand what you mean, I'll do that...

gpotter2 commented on test/run_tests:
```diff
@@ -7,7 +7,7 @@ else
 if [ -z "$*" ]
 then
-    PYTHONPATH=$DIR exec $PYTHON ${DIR}/scapy/tools/UTscapy.py -t regression.uts -f html -l -o /tmp/scapy_regression_test_$(date +%Y%m%d-%H%M%S).html
+    PYTHONPATH=$DIR exec $PYTHON ${DIR}/scapy/tools/UTscapy.py -t regression.uts -f html -l -o /tmp/scapy_regression_test_$(date +%Y%m%d-%H%M%S).html -K mock_read_routes6_windows
```
Yes: trying to load winpcapy under Linux will return an exception because it cannot load libpcap, and I didn't feel to add tons of crapy mocks to avoid loading it..

guedou replied:
That's now useless.

guedou commented:
Thanks, I understand. I wonder if we should keep architecture dependent tests outside of regression.uts as done with bpf.uts I might have done a mistake in the first place while adding the mocked read_routes6 tests.

That way there is no need to modify run_test
What do you think ?

gpotter2 commented:
Oh no... this requires the LOOPBACK_NAME changes -_-

@guedou so should we move all the read_routes6 to bpf.uts ?

p-l- commented:
Hi @gpotter2 #405 merged, can you rebase against master? Thanks!

gpotter2 commented:
@p-l- Should i squash all those commits in one ? 
Side note: plz delete my trashy builds...

p-l- commented:
I have a question for you @gpotter2 (maybe @guedou?): should we run `route_add_loopback(routes6, True)` when we first read the routes? So that we don't have to do so manually?

gpotter2 commented:
You mean in UTscapy ?

We could but we anyway as the regressions are flushing the routes, we at least would need to put it twice...

p-l- commented:
No, I mean in Scapy itself. Wouldn't it be useful to have the routes for the loopback interface anyway?

gpotter2 commented:
Oh understood, but we cannot do that:

On windows, by default, the "Microsoft Loopback adapter" is disabled... it does not appear anywhere and scapy cannot use it...

So we cannot presume that it is enabled: the `route_loopback_adapter` is a fake function that imports it for the tests only...

p-l- commented:
OK I got it. So what happens when you run `ping6 ::1` with the "Microsoft Loopback adapter" disabled? Does it work but you cannot sniff on it?

gpotter2 commented:
Exactly:
```
>>> sniff()
INFO: Sniffing on <NetworkInterface Killer E2200 Gigabit Ethernet Controller {C56DFFB3-992C-4964-B000-3E7C0F76E8BA}>
```

```
C:\Users\gpotter>ping -6 ::1

Pinging  ::1 with 32 bytes of data:
Reply from ::1 : time<1ms
Reply from ::1 : time<1ms
Reply from ::1 : time<1ms
Reply from ::1 : time<1ms
Reply from ::1 : time<1ms

Ping statistics for ::1:
    Packets: Sent = 5, Received = 5, Lost = 0 (0% loss)
Approximate round trip times in milli-seconds:
    Minimum = 0ms, Maximum = 0ms, Average = 0ms
```

p-l- commented:
So by default (sorry I know nothing about Windows or IPv6), what happens with scapy when you `(IPv6(dst="::1")/ICMPv6EchoRequest()).src`? Is the src IP address set to the interface with the default route?

gpotter2 commented:
Yes

p-l- commented:
And this is not what the IP stack would do, otherwise `ping6 ::1` would not work, am I rigth (sorry again for asking stupid questions)?

gpotter2 commented:
There's no stupid questions...

And yes, i think IP stack uses different stuff... (or maybe not, but internal windows program can sometimes use disabled functions)

p-l- commented:
Thanks for the clarification. I'll have to think about that / test by myself.

gpotter2 commented:
That was to fix the conflicts... (If you use the online "conflicts resolver" it calls it "Merge....")

gpotter2 commented:
As the code coverage integration is not complete, it's normal that some code isn't called...
Here, the "missing" code is called [here](https://github.com/secdev/scapy/pull/398/files#diff-38efe53471000211927337d537164033R36)

p-l- commented:
Can you rebase instead of using the online conflict fix (since it creates a merge commit)?

gpotter2 commented:
... I just destroyed my work by messing up the branch

gpotter2 commented:
@p-l- I rebased from secdev/master and squashed everything to make it looks better...

guedou commented on .travis/test.sh:
```diff
@@ -18,6 +18,8 @@ then
   UT_FLAGS+=" -K combined_modes"
 fi
 
+UT_FLAGS+=" -K mock_read_routes6_windows "
```
I think that its is better to keep `regression.uts` free from architecture dependent tests (it is not actually the case). It would be better to move your test into `mock_windows.uts`. We need to do something similar with the `read_routes6` ones.

What do you think ?

gpotter2 replied:
Good idea, where should we put the .uts ? I guess as it's part of the main scapy in test/ ?

guedou commented on appveyor.yml:
```diff
@@ -19,20 +19,20 @@ install:
   
   # Main unit tests
-  - "%PYTHON%\\python bin\\UTscapy -f text -t test\\regression.uts -F -K automaton -K mock_read_routes6_bsd || exit /b 42"
+  - "%PYTHON%\\python scapy\\tools\\UTscapy.py -f text -t test\\regression.uts -F -K automaton -K mock_read_routes6_bsd || exit /b 42"
```
do you need to change this ?

gpotter2 replied:
No sorry

guedou commented on scapy/arch/windows/__init__.py:
```diff
@@ -472,8 +474,76 @@ def read_routes_post2008():
+            except:
+                continue
+            scope = scapy.utils6.in6_getscope(match.group(2))
+            ret.append((match.group(2), scope, iface)) #(addr,scope,iface)
```
Could you add a space after your '#' ? There are several similar fixes to apply.

guedou commented on scapy/error.py:
```diff
@@ -60,6 +60,18 @@ def emit(self, record):
 log_interactive = logging.getLogger("scapy.interactive")  # logs in interactive functions
 log_loading = logging.getLogger("scapy.loading")          # logs when loading Scapy
 
+# 
```
useless.

gpotter2 replied:
obviously...

guedou commented on test/regression.uts:
```diff
@@ -5303,6 +5311,70 @@ ff02::%lo0/32                      ::1                            UC          -
 
 test_netbsd_7_0()
 
+= Windows
```
See my first comment: move these tests to `mock_windows.uts`

guedou commented on test/run_tests.bat:
```diff
@@ -3,7 +3,8 @@ set MYDIR=%cd%\..
 if [%1]==[] (
   SET date=%DATE%
-  python "%MYDIR%\scapy\tools\UTscapy.py" -t regression.uts -f html -o scapy_regression_test_%date:~6,4%_%date:~3,2%_%date:~0,2%.html
+  python "%MYDIR%\scapy\tools\UTscapy.py" -t regression.uts -K automaton -f html -o scapy_regression_test_%date:~6,4%_%date:~3,2%_%date:~0,2%.html
```
Do you think that automaton will ever work on Windows ?
If not we might want to move them to `automaton.uts` to make `regression.uts` architecture independent.

gpotter2 replied:
Well this is a problem we should consider, i do not understand why it does not currently work... I'll check that a bit...

gpotter2 commented:
@guedou I haven't looked at the automaton tests yet...

guedou commented:
@gpotter2 no problem. The PR looks good.

guedou commented on test/mock_windows.uts:
```diff
@@ -0,0 +1,70 @@
+# $Id: regression.uts,v 1.9 2008/07/29 15:30:29 pbi Exp pbi $
+
+############
+############
```
You need a title starting with "+". That's why there is an error with the appveror tests.

gpotter2 replied:
Thanks

gpotter2 commented:
You may of course squash this PR.

guedou commented on .travis/test.sh:
```diff
@@ -24,6 +24,7 @@ then
   UT_FLAGS+=" -K crypto "
 fi
 
+
```
Is this change mandatory ?

guedou commented on scapy/arch/windows/__init__.py:
```diff
@@ -500,22 +569,43 @@ def get_working_if():
-
-def route_add_loopback():
+def route_add_loopback(routes=None, ipv6=False, iflist=None):
     """Add a route to 127.0.0.1 to simplify unit tests on Windows"""
```
change with: Add a route to 127.0.0.1 and ::1 ....

guedou commented on test/mock_windows.uts:
```diff
@@ -0,0 +1,72 @@
+% Regression tests on Windows only for Scapy
+
+# More informations at http://www.secdev.org/projects/UTscapy/
+# $Id: regression.uts,v 1.9 2008/07/29 15:30:29 pbi Exp pbi $
```
You can delete this line.

guedou commented on test/regression.uts:
```diff
@@ -1,7 +1,7 @@
 
 # More informations at http://www.secdev.org/projects/UTscapy/
-# $Id: regression.uts,v 1.9 2008/07/29 15:30:29 pbi Exp pbi $
+# $Id: regression.uts,v 1.9.1 2016/05/01 22:58:12 pbi Exp pbi $
```
You could also remove this one, as it is/wa a CVS generated string.

p-l- commented:
BTW, you have a merge commit ;-)

gpotter2 commented:
@p-l- I know, but I needed this one in order to modify appveyor (without conflicts)

p-l- commented:
@gpotter2: `git checkout master; git pull; git checkout gpotter2-ipv6; git rebase -i master`?

gpotter2 commented:
@p-l- I couldn't because there was a conflict (i modified appveyor.yml on this branch, then it was modified on secdev/master, then i modified it again), so i just squashed every thing...

gpotter2 commented:
Oh before https://github.com/secdev/scapy/pull/452 this bug was hidden... It's due to the us end of the fake interface... I'll fix it

guedou commented:
@p-l good for me. I still don't like the `-K automaton ` in run_test.bat but that the best/easiest option today.

p-l- commented:
@gpotter2 that's the whole purpose of rebase-ing: you can fix the conflicts easily.

p-l- commented:
Great work @gpotter2, thanks! Thanks also @guedou!
